ports/machine_pwm: Set the PWM duty_u16() back to 65535.

robert-hh opened the pull request:
Tested that output is high at duty_u16(65565) and low at duty_u16(0). Also verified that at duty_u16(32768) the high and low pulse have the same length.

Tested with ESP8266, MIMXRT, NRF and SAMD, which are affected by this change. Other ports did not require a change.

Partially reverting PR #10850.

robert-hh committed:
ports: Make PWM duty_u16 have an upper value of 65535 across all ports.

The following ports used 65536 as the upper value (100% duty cycle) and are
changed in this commit to use 65535: esp8266, mimxrt, nrf, samd.

Tested that output is high at `duty_u16(65535)` and low at `duty_u16(0)`.
Also verified that at `duty_u16(32768)` the high and low pulse have the
same length.

Partially reverts #10850, commits 9c7ad68165bcd224c94ca6d8f172362cf8000d99
and 2ac643c15bec8c88ece0e944ce58f36d02dfd2dd.

Signed-off-by: robert-hh <robert@hammelrath.com>

dpgeorge commented on ports/mimxrt/hal/pwm_backport.c:
```diff
@@ -36,15 +36,15 @@ void PWM_UpdatePwmDutycycle_u16(
     // Setup the PWM dutycycle of channel A or B
     if (pwmSignal == kPWM_PwmA) {
-        if (dutyCycle >= 65536) {
+        if (dutyCycle >= PWM_FULL_SCALE) {
```
This constant is still defined as 65536 in `pwm_backport.h`.  Does it need changing there?

robert-hh replied:
Of course.

robert-hh replied:
Updated. Behavior still fine. At duty_u16(32768) the difference between low and high phases with a Teensy 4.x are 700ps (flexpwm) and 400ps (QTMR).

dpgeorge replied:
Can the hardware have even better symmetry with 50% duty cycle, if you choose the right upper limit for the counter?  If so, it may be possible to get this working with `machine.PWM` by using 65536 for the hardware full scale, but 65535 for the `duty_u16` max, and then scaling it via `x * 65536 / 65535`.  I think this is how the rp2 port works.

robert-hh replied:
The difference between the high/low phase is 10% of the PWM clock or about 50% of the CPU clock. It hardly can get much better.

dpgeorge replied:
Right... picoseconds are very small!  Great work testing it to that degree.

projectgus reviewed (commented):
This looks good to me, thanks @robert-hh

mattytrentini commented:
Thanks for taking care of all of these important details folks!

robert-hh commented:
1. A suitable pin pair on Teensy 4.x wich uses QTMR and has a UART is D14/D15 and UART3.
2. The PWM test passes with FLEXPWM_AB pins if duty and freq are set in a single call using pwm.init() and not  separate. That looks at first suprising, since all methods of setting PWM parameters use the same function for actually configuring the hardware and the sequence should not matter. I tested that initially, but not with the sequence duty_u16(65535) -> freq(x) -> duty_u16(0). That's the one which make trouble.

robert-hh committed:
mimxrt/machine_pwm: Fix a few inconsistencies with PWM output.

Changes in this commit:
- When setting PWM parameters of a FLEXPWM AB channel twice within a PWM
  cycle, the second setting was ignored. Now the second setting persists.
- With `duty_u16(0)` a FLEXPWM X channel was set to high impedance.  Now it
  is set to low impedance with value 0 for `invert=False`, and 1 for
  `invert=True`.
- The align parameter requires a duty rate and frequency to be set.  Align
  will now be ignored if freq or duty are missing.

Signed-off-by: robert-hh <robert@hammelrath.com>

robert-hh commented:
The issues with the FLEXPM channels and duty_u16(0) are fixed. The changes are pushed. Tested with MIMXRT 1011, -1015, -1020, -1052, -1062 (Teensy 4.x) and -1176.

dpgeorge commented:
Excellent work @robert-hh !  We should probably make more of these hardware tests, they are really teasing out the bugs.

> The MIMXRT port has three different mechanisms for PWM

I think we should make the PWM test test all three of these mechanisms (for mimxrt boards).  Can you suggest three pairs of pins that can be used for this?  Maybe one pair can also have MISO/MOSI to test loopback SPI.

robert-hh commented:
>  Can you suggest three pairs of pins that can be used for this? Maybe one pair can also have MISO/MOSI to test loopback SPI.

Sure:
Pin D0/D1: FLEXPMW X and UART 1
Pin D2/D3: FLEXPWM A/B
Pin D11/D12: QTMR and MOSI/MISO of SPI 0

dpgeorge commented:
Merged.  Thanks again @robert-hh for your efforts here.

robert-hh commented:
Thank you for setting up the test, which proves to be very useful.
